Viewer camera improvements

johnhaddon opened the pull request:
The main feature request that motivated this was to have a way of automatically fitting the viewer's clipping planes to the current scene, but I've also improved the camera section of the main toolbar based on previous comments.

@andrewkaufman, I expect we'll want to back port this one to the 0.3X branch. That should be straightforward enough, but note that we'll also need to add a fix for the binding of `SceneGadget::getSelection()`. This is broken on 0.3X and was fixed recently on master by c3a935951fbdfb7f4742b3f69504f779b089a786, but we'll probably need a different approach for the backport fix.

andrewkaufman commented:
I like the camera icon and most of the new behaviour, but playing with this interactively, the fit menu items feel a bit odd.

- Fit to Selection sounds like it would give the minimal clipping for the current selection, but in practice (on both a manufactured scene and a full scale production scene) it feels like its showing me too much.
  - For example
    - I selected a single building in a city and Fit to Selection, but my clipping planes are set (0.1, 10000).
    - I then framed to that building, and Fit to Selection again, just in case the current view was influencing the fit, but got a similar result.
    - Playing with the planes myself, farClip of 3500 seems sufficient to show the entire building, so I guess the log10 is causing me to see more than I want. Do we need that? Is there a tighter "make the numbers nice" metric we can use instead?
  - Maybe we need a more explicit "fit exactly this bound without padding for tumble room"?

- Is ExpandAll respecting the clipping planes in terms of drawing? Tumbling performance on my city is quite slow with ExpandAll, even with very tight clipping planes.

- The fit options aren't very discoverable. Is this the first context menu in the scene viewer? I sort of expected to find them on the settings menu as context options on the clipping planes plug. Or perhaps on the camera dropdown itself. I don't mind them in the viewer, but I'm not sure anyone will find them there.

- Should we have a shortcut for Fit to Scene and Reset? Maybe other modifiers on K? Should Reset be called Revert to Default like it is in other parts of the UI?

Aside from fit, I noted these other issues:

- Look through a camera and then switch nodes to a scene that doesn't include that camera. You're left in a confusing locked state with no real indication of what went wrong or what you're trying to look through.

- Now that its so easy to monkey with the default camera, it feels unintuitive that we can't yet monkey with the real cameras in the same way. Since we've planned that part of it, I assume you're intending followup PRs that add those features. Maybe its worth holding back this from a release until those are ready (e.g pushing out 0.45 before merging this)?

johnhaddon commented:
> Fit to Selection sounds like it would give the minimal clipping for the current selection, but in practice (on both a manufactured scene and a full scale production scene) it feels like its showing me too much.  I guess the log10 is causing me to see more than I want. Do we need that? Is there a tighter "make the numbers nice" metric we can use instead?

It's up for debate definitely, and easy to change. I'm not totally happy with it. I based my current choice on a couple of things :

- I don't think users want a tight fit. I think they want to set the clipping planes once for their scene and then never have to worry about it again, even if they tumble, zoom out etc.
- On Friday Alex showed me a simple manual tool they use that just provides a context menu containing a set of increasing power-of-ten ranges to choose from. That made me think that power-of-ten ranges were neat and sufficient.

> Maybe we need a more explicit "fit exactly this bound without padding for tumble room"?

If that's something users need, then no problem. I'm aiming for a minimal set of useful options though, so if it's not needed then we should omit it. I'm working on the assumption they want to be in the right ballpark and not to touch it again, rather than to adjust regularly as they more around.

> Is ExpandAll respecting the clipping planes in terms of drawing? Tumbling performance on my city is quite slow with ExpandAll, even with very tight clipping planes.

ExpandAll controls what we do in the threaded evaluation of the scene, prior to making any GL calls, and changing that would be fairly non-trivial I think. We could improve tumbling performance by doing basic frustum culling though (including the clipping planes) - it's kindof embarrassing that we don't do that already.

> The fit options aren't very discoverable. Is this the first context menu in the scene viewer?

It is. I expect we'll add lots more now we can though - one obvious one is to quickly open a SceneInspector history traceback for the selected object. I hope this menu will become an obvious "go to" place - we rely on context menus in all sorts of other places after all. 

> I sort of expected to find them on the settings menu as context options on the clipping planes plug. Or perhaps on the camera dropdown itself. I don't mind them in the viewer, but I'm not sure anyone will find them there.

I think they belong in the viewer for sure, because they advertise a shortcut which only works in the viewer. Adding them to the plug in the settings window would be fine too, although I don't think that's much more discoverable, and we won't be adding all sorts of other cool stuff there in the future. I definitely don't think they should be in the camera dropdown itself - this is fairly esoteric compared to the high-level operations it provides.

> Should we have a shortcut for Fit to Scene and Reset? Maybe other modifiers on K?

I did think about it and decided against it. Framing is a far more common and important operation, and nobody has ever asked for modifiers on F. The current fit shortcut actually works like framing in that if there's a selection it fits to it, and if there isn't, it fits to the scene. 

> Should Reset be called Revert to Default like it is in other parts of the UI?

Yeah, I'd buy that, especially if we also add the same menu to the plug context menu in the Camera Settings window.

> Look through a camera and then switch nodes to a scene that doesn't include that camera. You're left in a confusing locked state with no real indication of what went wrong or what you're trying to look through.

Yeah, that has always been confusing. Maybe we keep the resolution gate up and put an error message alongside the camera name?

> Now that its so easy to monkey with the default camera, it feels unintuitive that we can't yet monkey with the real cameras in the same way. Since we've planned that part of it, I assume you're intending followup PRs that add those features. Maybe its worth holding back this from a release until those are ready (e.g pushing out 0.45 before merging this)?

Are you talking about the camera movement tool? I've done a bunch of parallel work on that but I'm not sure how close I am to having a PR. My understanding was that the clipping plane stuff was quite urgent for one set of users, and the camera movement was slightly less urgent for a different set of users. If that's correct, I'm inclined to get this out rather than hold it up waiting for the camera tool.

andrewkaufman commented:
> It's up for debate definitely, and easy to change. I'm not totally happy with it

Ok, I'll try to gather some user feedback at my desk then.

> We could improve tumbling performance by doing basic frustum culling though (including the clipping planes) - it's kindof embarrassing that we don't do that already.

Indeed. I guess that should go on my list of 'viewer performance tickets' that still need creating.

> Adding them to the plug in the settings window would be fine too

Ok, lets have em in both spots then.

>  The current fit shortcut actually works like framing in that if there's a selection it fits to it, and if there isn't, it fits to the scene.

Oh great. I did not realize that.

>> Should Reset be called Revert to Default like it is in other parts of the UI?
> Yeah, I'd buy that

Its a deal.

>  Maybe we keep the resolution gate up and put an error message alongside the camera name?

That's what I was thinking, maybe with a red cross through the gate or something? It used to be somewhat more noticable because you'd be confused, but eventually glance at the LookThrough path and realize what's going on. With the icon, you've got no indication which camera it was supposed to be.

> Are you talking about the camera movement tool? 

Yeah. Now that the mockup is in the clipping plane ticket, I'm not sure we should take the pressure off until its done. I can backport just the clipping bit, but I was meaning more for 0.46 that they should both come out together.

andrewkaufman commented:
> I was meaning more for 0.46 that they should both come out together.

Ehh, scratch that. Lets just get this little fixes in and merge. 0.45 is "alpha" anyways.

andrewkaufman commented:
Notes from @murrays-ie playing with it at my desk:

- Agree the rounding is too loose. Can we ceil to next significant digit rather than order of magnitude?
- Multiple Viewers and multiple Camera Settings windows : Which goes with which?
- Messing with FOV and not re-focusing makes tumbling awkward. Can the plug trigger an adjustment somehow?
- Can we also have Ctrl+F to frame + set clipping planes? It makes for a smoother workflow when framing and realizing you need to adjust the planes.
- Setting clipping planes doesn't seem esoteric. Why shouldn't it be a first order UI element (ie. in the camera dropdown)?
- Can we have a visual indication on the camera icon that its in use (e.g. not set to Default)? Kind of like the icons on the image viewer.

johnhaddon commented:
Pushed some updates with where I'm at :

> - Put warning on resolution gate when camera doesn't exist

Done in 1601b8ea656cd65a8cdfdac677f6fb876765ecab.

> Can we have a visual indication on the camera icon that its in use (e.g. not set to Default)? Kind of like the icons on the image viewer.

Nice suggestion. 288f2a3d44dcc01034e187d6612cbd139e16d1d1.

> Put the fit menu items in the settings window too

Done in 1d7fd2de81a356d567c85c7d7c86e470b172dcfc. Note I've not renamed to "Revert to Defaults" as suggested because that isn't actually the name used for plug popup menus. I've used "Default" so they both match, which I think was the intention anyway.

>  Can we also have Ctrl+F to frame + set clipping planes? It makes for a smoother workflow when framing and realizing you need to adjust the planes.

4b769b4d9551118b7084ba1b085e62cc5b9ea3f5. Note that this required changes to the View base class, some of which you shouldn't backport. I've split that into a [separate commit](c7d0bee).

> - Agree the rounding is too loose. Can we ceil to next significant digit rather than order of magnitude?

I may be being dumb, but I'm not quite sure what you mean by this. Could you elaborate?

> - Multiple Viewers and multiple Camera Settings windows : Which goes with which?

I don't have any bright ideas here, do you? The Viewer is identified by a title in the tab, but that's not unique. I would hope that generally you don't need to keep those settings open for long enough for it to become a big issue...

> - Setting clipping planes doesn't seem esoteric. Why shouldn't it be a first order UI element (ie. in the camera dropdown)?

The way I see it, the camera dropdown's main purpose is to choose a camera - a first order operation. Filling it with lesser operations makes this less clear. Framing is a far more common operation and it isn't in there, nor has anyone ever requested that. If clipping options belong there, then so could any number of other things, and we'll just end up with clutter. There are already two places to set clipping, and two separate shortcuts for it. That's plenty.

> Messing with FOV and not re-focusing makes tumbling awkward. Can the plug trigger an adjustment somehow?

I think I know what this is. We lose the centre of interest when switching cameras - the problem already existed when switching between the default and look-through cameras. I did try to fix this yesterday in fact, but hit some nasty side effects. Will revisit tomorrow.

Cheers...
John

andrewkaufman commented:
>> Can we ceil to next significant digit rather than order of magnitude?

> I may be being dumb, but I'm not quite sure what you mean by this. Could you elaborate?

I think it was a (mental, rather than tried and tested) stab at making the values tighter while still keeping some notion of "nice numbers" that the commit is [mentioning](https://github.com/johnhaddon/gaffer/blob/989b9869f91bd7c51b081c98bc2352c9b01ceda7/src/GafferUI/ViewportGadget.cpp#L692).

I just tried (from the ScriptEditor) using the building I was trying to focus on before, and this calc gives me clipping planes that are pretty close to what I would have chosen manually:

```
niceFar = round( far, -(farLogTen-1) )
imath.V2f( pow( 10, nearLogTen ), niceFar )
```

>> Multiple Viewers and multiple Camera Settings windows : Which goes with which?

> I don't have any bright ideas here, do you? 

I guess naming the viewers (e.g Viewer1 : SceneReader, Viewer2: Sphere)? Lets leave it for another time...

danieldresser-ie commented:
Hmm, looks like there isn't any real standard approach to "Round-To-N-Significant-Digits" in C++.

I was thinking of something like this ( same result as Andrew's code, but without using that Python round() function ):
```
if( far != 0 )
{
  farLogTen = ceil( log10( abs( far ) ) );
  farMagnitude = pow( 10, farLogTen - 1)
  far = ceil( far / farMagnitude ) * farMagnitude
}
```

johnhaddon commented:
OK, I've rebased and addressed the two outstanding items.

> Messing with FOV and not re-focusing makes tumbling awkward. Can the plug trigger an adjustment somehow?

This turned out to be crazy simple compared to what I'd been doing yesterday - 84a5710. I may need to return to a more complex approach to support the camera tool, but I'll deal with that separately.

> Can we ceil to next significant digit rather than order of magnitude?

In b5080f1607587f85a4db8a62888659ce2087a7c4 I've taken Daniel's code for the far plane, and made something up that I thought was sensible for the near plane. Can take another stab at it if this isn't right, but hopefully it doesn't seem totally unreasonable...

andrewkaufman commented:
Looks great, thanks!
